File: impl/csharp/test/ChannelsKatTest.cs
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;

using Naalp;
using Xunit;

namespace Sh.Bubblefish.Naalp.Tests
{
    /// <summary>
    /// The C10 channels wave (task #174) for the C# SDK: the <see cref="Channels.AllowedTransition"/>
    /// state-machine guard, the public <see cref="Channels.ChannelSpec"/>/<see cref="Channels.GetChannel"/>
    /// parity accessor, <see cref="Channels.CheckFrameTree"/> (Spatial TransformCycle, §20), and the
    /// durable WAL-backed <see cref="Channels.WorkflowGate"/> whose crash test proves InputGateBypass
    /// cannot occur (design-channels.md §18).
    ///
    /// <para>Graded against the independent, non-circular <c>vectors/channels/&lt;name&gt;/cases.json</c>
    /// corpus (NOT produced by this code) — the same corpus impl/go/channels/channels_test.go
    /// (TestTableMatchesOracle, TestStateMachine, TestTransformCycle, TestWorkflowInputGateBypass) and
    /// impl/rust/src/channels.rs's <c>#[cfg(test)]</c> module grade against. There is no CBOR/wire
    /// encoding in this surface (the twenty-channel registry and the WAL gate are internal state, not
    /// signed objects) — matching Go/Rust, which likewise define no encoder here — so this KAT grades
    /// verdicts (allowed/denied, accept/reject, gate status) rather than byte parity.</para>
    ///
    /// <para>Run: <c>dotnet test test/Bubblefish.Naalp.Tests.csproj --filter FullyQualifiedName~ChannelsKatTest</c></para>
    /// </summary>
    public sealed class ChannelsKatTest
    {
        // The twenty channel names in Table order (design-channels.md §1..§20), lower-cased to match
        // the vectors/channels/<name>/ directory naming.
        private static readonly string[] AllChannelNames =
        {
            "control", "memory", "capability", "identity", "governance", "immune", "federation",
            "settlement", "compliance", "sensory", "telemetry", "audit", "stream", "bridge",
            "commerce", "interaction", "discovery", "workflow", "knowledge", "spatial",
        };

        private static JsonElement LoadChannel(string dir)
        {
            var d = new DirectoryInfo(AppContext.BaseDirectory);
            string rel = Path.Combine("vectors", "channels", dir, "cases.json");
            for (int i = 0; i < 12 && d != null; i++)
            {
                string p = Path.Combine(d.FullName, rel);
                if (File.Exists(p))
                {
                    return JsonDocument.Parse(File.ReadAllText(p)).RootElement.Clone();
                }
                d = d.Parent;
            }
            throw new Xunit.Sdk.XunitException(rel + " not found from " + AppContext.BaseDirectory);
        }

        // ==== full descriptor parity: id/name/kinds/states/transitions/errors == the independent
        //      oracle, for all twenty channels (mirrors Go TestTableMatchesOracle / Rust table_matches_oracle) =

        [Fact]
        public void ChannelSpecMatchesOracleForAllTwentyChannels()
        {
            int n = 0;
            foreach (string dir in AllChannelNames)
            {
                JsonElement c = LoadChannel(dir);
                long id = c.GetProperty("channel_id").GetInt64();
                Channels.ChannelSpec? spec = Channels.GetChannel(id);
                Assert.NotNull(spec);
                Channels.ChannelSpec s = spec!;
                Assert.Equal(id, s.Id);
                Assert.Equal(c.GetProperty("name").GetString(), s.Name);

                JsonElement kinds = c.GetProperty("kinds");
                Assert.Equal(kinds.GetArrayLength(), s.Kinds.Length);
                int ki = 0;
                foreach (JsonElement k in kinds.EnumerateArray())
                {
                    Channels.KindSpec got = s.Kinds[ki];
                    Assert.True(k.GetProperty("code").GetInt64() == got.Code, dir + " kind " + ki + " code");
                    Assert.True(k.GetProperty("name").GetString() == got.Name, dir + " kind " + ki + " name");
                    Assert.True(k.GetProperty("effect").GetInt64() == got.Effect, dir + " kind " + ki + " effect");
                    Assert.True(k.GetProperty("variable").GetBoolean() == got.Variable, dir + " kind " + ki + " variable");
                    ki++;
                }

                JsonElement states = c.GetProperty("states");
                Assert.Equal(states.GetArrayLength(), s.States.Length);
                int si = 0;
                foreach (JsonElement st in states.EnumerateArray())
                {
                    Assert.True(st.GetString() == s.States[si], dir + " state " + si);
                    si++;
                }

                JsonElement transitions = c.GetProperty("transitions");
                Assert.Equal(transitions.GetArrayLength(), s.Transitions.Length);
                int ti = 0;
                foreach (JsonElement t in transitions.EnumerateArray())
                {
                    Assert.True(t.GetProperty("from").GetString() == s.Transitions[ti].From, dir + " transition " + ti + " from");
                    Assert.True(t.GetProperty("to").GetString() == s.Transitions[ti].To, dir + " transition " + ti + " to");
                    ti++;
                }

                JsonElement errors = c.GetProperty("errors");
                Assert.Equal(errors.GetArrayLength(), s.Errors.Length);
                int ei = 0;
                foreach (JsonElement e in errors.EnumerateArray())
                {
                    Assert.True(e.GetString() == s.Errors[ei], dir + " error " + ei);
                    ei++;
                }
                n++;
            }
            Assert.Equal(20, n);
        }

        [Fact]
        public void GetChannelReturnsNullForUnregisteredId()
        {
            Assert.Null(Channels.GetChannel(0x00FF));
            Channels.ChannelSpec? wf = Channels.GetChannel(0x0011);
            Assert.NotNull(wf);
            Assert.Equal("Workflow", wf!.Name);
        }

        // ==== AllowedTransition: every declared transition (per channel, read from the independent
        //      oracle) is allowed; representative reversed/skip/unregistered pairs are denied ==========

        [Fact]
        public void AllowedTransitionAcceptsEveryDeclaredTransitionAcrossAllChannels()
        {
            int checkedCount = 0;
            foreach (string dir in AllChannelNames)
            {
                JsonElement c = LoadChannel(dir);
                long id = c.GetProperty("channel_id").GetInt64();
                foreach (JsonElement t in c.GetProperty("transitions").EnumerateArray())
                {
                    string from = t.GetProperty("from").GetString()!;
                    string to = t.GetProperty("to").GetString()!;
                    Assert.True(Channels.AllowedTransition(id, from, to), dir + ": " + from + "->" + to + " must be allowed");
                    checkedCount++;
                }
            }
            Assert.True(checkedCount > 0, "no transitions loaded from the oracle");
        }

        [Fact]
        public void AllowedTransitionRejectsReversedSkipAndUnregistered()
        {
            // Mirrors impl/go/channels/channels_test.go TestStateMachine and
            // impl/rust/src/channels.rs state_machine byte-for-byte on the verdict.
            Assert.True(Channels.AllowedTransition(0x0001, "offered", "accepted"));   // Memory
            Assert.True(Channels.AllowedTransition(0x0001, "live", "revoked"));       // Memory
            Assert.False(Channels.AllowedTransition(0x0001, "revoked", "live"));      // Memory regression
            Assert.True(Channels.AllowedTransition(0x000E, "order", "fulfil"));       // Commerce
            Assert.False(Channels.AllowedTransition(0x000E, "offer", "fulfil"));      // Commerce skip
            Assert.True(Channels.AllowedTransition(0x0011, "awaiting-input", "running"));   // Workflow
            Assert.False(Channels.AllowedTransition(0x0011, "created", "running"));         // Workflow gate skip
            Assert.False(Channels.AllowedTransition(0x0001, "accepted", "offered"));  // reversed, never declared
            Assert.False(Channels.AllowedTransition(0x00FF, "a", "b"));               // unregistered channel
        }

        // ==== CheckFrameTree: Spatial's TransformCycle (design-channels.md §20) ======================

        [Fact]
        public void CheckFrameTreeAcceptsValidRejectsCyclic()
        {
            var tree = new Dictionary<string, string> { ["base"] = "", ["arm"] = "base", ["hand"] = "arm" };
            Channels.CheckFrameTree(tree); // must not throw

            var cyclic = new Dictionary<string, string> { ["a"] = "b", ["b"] = "c", ["c"] = "a" };
            NaalpException ex = Assert.Throws<NaalpException>(() => Channels.CheckFrameTree(cyclic));
            Assert.Equal("TransformCycle", ex.Kind);
        }

        [Fact]
        public void CheckFrameTreeAcceptsForestOfIndependentRoots()
        {
            // Two disjoint single-node roots form a valid (degenerate) forest -- not a cycle.
            var singles = new Dictionary<string, string> { ["x"] = "", ["y"] = "" };
            Channels.CheckFrameTree(singles); // must not throw
        }

        // ==== WorkflowGate: the durable input-gate crash test (design-channels.md §18) ===============
        //
        // MUTATION ANCHOR: Channels.WorkflowGate.Run's awaiting-input/awaiting-approval branch (which
        // throws InputGateBypass) is what WorkflowInputGateBypassMirrorsGoCrash exercises. Neutering
        // Run to persist "running" unconditionally -- skipping that branch -- flips this test from
        // pass to fail (recorded red evidence).

        private static string TempGatePath(string tag) =>
            Path.Combine(Path.GetTempPath(), "naalp-wf-gate-" + tag + "-" + Guid.NewGuid().ToString("N") + ".wal");

        [Fact]
        public void WorkflowInputGateBypassMirrorsGoCrash()
        {
            string path = TempGatePath("bypass");
            try
            {
                Channels.WorkflowGate g = Channels.OpenWorkflowGate(path);
                try
                {
                    g.Create("t1", false);

                    // Running before input is InputGateBypass.
                    NaalpException ex1 = Assert.Throws<NaalpException>(() => g.Run("t1"));
                    Assert.Equal("InputGateBypass", ex1.Kind);
                }
                finally
                {
                    // Simulate a crash right after Create: close the WAL (mirrors Go's explicit
                    // g.Close() in the same test).
                    g.Close();
                }

                Channels.WorkflowGate g2 = Channels.OpenWorkflowGate(path);
                try
                {
                    Assert.Equal("awaiting-input", g2.Status("t1")); // crash did not bypass the gate

                    NaalpException ex2 = Assert.Throws<NaalpException>(() => g2.Run("t1"));
                    Assert.Equal("InputGateBypass", ex2.Kind);

                    // Supplying input opens the gate; only then may it run.
                    g2.SupplyInput("t1");
                    g2.Run("t1"); // must not throw
                    Assert.Equal("running", g2.Status("t1"));
                }
                finally
                {
                    g2.Close();
                }
            }
            finally
            {
                if (File.Exists(path)) File.Delete(path);
            }
        }

        [Fact]
        public void WorkflowGateApprovalPathAndTaskStateErrors()
        {
            string path = TempGatePath("approval");
            try
            {
                Channels.WorkflowGate g = Channels.OpenWorkflowGate(path);
                try
                {
                    // needsApproval path: created -> awaiting-approval -> approved -> running.
                    g.Create("t-appr", true);
                    Assert.Equal("awaiting-approval", g.Status("t-appr"));
                    NaalpException gateEx = Assert.Throws<NaalpException>(() => g.Run("t-appr"));
                    Assert.Equal("InputGateBypass", gateEx.Kind);
                    g.SupplyInput("t-appr");
                    Assert.Equal("approved", g.Status("t-appr"));
                    g.Run("t-appr");
                    Assert.Equal("running", g.Status("t-appr"));

                    // Duplicate Create on an existing task id is TaskStateError, not silently accepted.
                    NaalpException dupEx = Assert.Throws<NaalpException>(() => g.Create("t-appr", false));
                    Assert.Equal("TaskStateError", dupEx.Kind);

                    // SupplyInput/Run on an unknown task id is TaskStateError, never InputGateBypass --
                    // there is no gate to bypass when the task does not exist.
                    NaalpException unkSupply = Assert.Throws<NaalpException>(() => g.SupplyInput("ghost"));
                    Assert.Equal("TaskStateError", unkSupply.Kind);
                    NaalpException unkRun = Assert.Throws<NaalpException>(() => g.Run("ghost"));
                    Assert.Equal("TaskStateError", unkRun.Kind);

                    // Supplying input twice (already past the gate) is TaskStateError.
                    NaalpException dupSupply = Assert.Throws<NaalpException>(() => g.SupplyInput("t-appr"));
                    Assert.Equal("TaskStateError", dupSupply.Kind);

                    Assert.Null(g.Status("never-created"));
                }
                finally
                {
                    g.Close();
                }
            }
            finally
            {
                if (File.Exists(path)) File.Delete(path);
            }
        }
    }
}

File: impl/kotlin/src/test/kotlin/ChannelsKatTest.kt
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
package sh.bubblefish.naalp

import java.io.File

/**
 * C10 channel-state-machine + Workflow-input-gate cluster (design-channels.md §1..§20, §18)
 * known-answer + fail-closed tests for the Kotlin SDK.
 *
 * CORPUS-GRADED (pure, non-circular per F3): every one of the twenty channels' (channel_id, name,
 * kinds[code/name/effect/variable], states, transitions[from/to], errors) against the independent
 * per-channel oracle under vectors/channels/<lowercase-name>/cases.json — the SAME oracle Go and
 * Rust grade against (Kotlin == Go == Rust == oracle on the whole channel table).
 *
 * ISOLATION (self-contained, no vector — behavioural, mirrors Go channels_test.go /
 * impl/rust/src/channels.rs #[cfg(test)]): AllowedTransition representative accept/reject cases;
 * CheckFrameTree accept-valid-tree / reject-cyclic; and the Workflow durable input/approval gate
 * crash test (design-channels.md §18) that proves InputGateBypass cannot occur even across a
 * simulated crash (close -> reopen with no input yet supplied).
 *
 * The load-bearing MUTATION ANCHOR: WorkflowGate.run's `"awaiting-input", "awaiting-approval" ->
 * throw InputGateBypass` branch is the fail-closed check that a task may not execute on input that
 * was never supplied/authorized. Neutering it (letting an ungated task fall through to "running")
 * flips the "task cannot run before its input gate" / "task cannot run after a crash without
 * passing the gate" assertions in workflowGateInputBypassRun() below from PASS to FAIL.
 *
 * KAT convention: a standalone main() exiting non-zero on any failure (mirrors ApprovalKatTest.kt /
 * RotationKatTest.kt / ProducingBoundaryKatTest.kt); the filename carries the "KatTest" token the
 * ten-language-parity gate indexes.
 */

// ---- balanced-brace / regex JSON access (string-aware; mirrors ApprovalKatTest.kt) ----------------

private class ChanFailure(msg: String) : RuntimeException(msg)

private var katFails = 0

private fun katOk(name: String) = println("  ok   $name")

private fun katFail(name: String, detail: String) {
    katFails++
    println("  FAIL $name\n       $detail")
}

private fun katCheck(name: String, got: String, want: String) {
    if (got == want) katOk(name) else katFail(name, "got  $got\n       want $want")
}

private fun katCheckBool(name: String, got: Boolean, want: Boolean) {
    if (got == want) katOk(name) else katFail(name, "got $got want $want")
}

/** Run [block]; require it throws a [NaalpException] whose kind == [wantKind]. */
private fun katExpectThrows(name: String, wantKind: String, block: () -> Unit) {
    try {
        block()
        katFail(name, "expected $wantKind, no exception thrown")
    } catch (e: NaalpException) {
        katCheck(name, e.kind, wantKind)
    }
}

/** Run [block]; require it does NOT throw. */
private fun katExpectOk(name: String, block: () -> Unit) {
    try {
        block()
        katOk(name)
    } catch (e: NaalpException) {
        katFail(name, "expected no error, got ${e.kind}: ${e.message}")
    }
}

private fun bracketExtent(s: String, openCh: Char, closeCh: Char, start: Int): Int {
    var depth = 0
    var inString = false
    var escape = false
    for (i in start until s.length) {
        val c = s[i]
        if (inString) {
            when {
                escape -> escape = false
                c == '\\' -> escape = true
                c == '"' -> inString = false
            }
            continue
        }
        when {
            c == '"' -> inString = true
            c == openCh -> depth++
            c == closeCh -> {
                depth--
                if (depth == 0) return i
            }
        }
    }
    return -1
}

/** The raw substring (including delimiters) of the value following "key": in [s], for an object
 *  ("{...}") or array ("[...]") value. */
private fun rawValue(s: String, key: String): String? {
    val idx = s.indexOf("\"$key\"")
    if (idx < 0) return null
    val after = s.substring(idx)
    val colon = after.indexOf(':')
    if (colon < 0) return null
    val rest = after.substring(colon + 1).trimStart()
    if (rest.isEmpty()) return null
    val opener = rest[0]
    val closer = when (opener) {
        '{' -> '}'
        '[' -> ']'
        else -> return null
    }
    val end = bracketExtent(rest, opener, closer, 0)
    if (end < 0) return null
    return rest.substring(0, end + 1)
}

private fun objectField(s: String, key: String): String? = rawValue(s, key)

private fun splitJsonObjects(body: String): List<String> {
    val out = ArrayList<String>()
    var depth = 0
    var start = -1
    var inString = false
    var escape = false
    for (i in body.indices) {
        val c = body[i]
        if (inString) {
            when {
                escape -> escape = false
                c == '\\' -> escape = true
                c == '"' -> inString = false
            }
            continue
        }
        when (c) {
            '"' -> inString = true
            '{' -> {
                if (depth == 0) start = i
                depth++
            }
            '}' -> {
                depth--
                if (depth == 0 && start >= 0) {
                    out.add(body.substring(start, i + 1))
                    start = -1
                }
            }
        }
    }
    return out
}

/** The list of raw JSON object strings inside "key": [ {...}, {...}, ... ]. */
private fun objectArrayField(s: String, key: String): List<String> {
    val raw = rawValue(s, key) ?: return emptyList()
    if (!raw.startsWith("[")) return emptyList()
    return splitJsonObjects(raw.substring(1, raw.length - 1))
}

/** The list of plain quoted strings inside "key": [ "a", "b", ... ]. */
private fun stringArrayField(s: String, key: String): List<String> {
    val raw = rawValue(s, key) ?: return emptyList()
    if (!raw.startsWith("[")) return emptyList()
    val body = raw.substring(1, raw.length - 1)
    return Regex("\"((?:[^\"\\\\]|\\\\.)*)\"").findAll(body).map { it.groupValues[1] }.toList()
}

private fun nullableField(s: String, key: String): String? {
    val m = Regex("\"" + Regex.escape(key) + "\"\\s*:\\s*(null|\"[^\"]*\"|-?[0-9]+|true|false)").find(s) ?: return null
    val raw = m.groupValues[1]
    if (raw == "null") return null
    return if (raw.startsWith("\"")) raw.substring(1, raw.length - 1) else raw
}

private fun strField(s: String, key: String): String =
    nullableField(s, key) ?: throw ChanFailure("string key not found: $key")

private fun intField(s: String, key: String): Long =
    nullableField(s, key)?.toLong() ?: throw ChanFailure("int key not found: $key")

private fun boolField(s: String, key: String): Boolean =
    nullableField(s, key)?.toBoolean() ?: throw ChanFailure("bool key not found: $key")

private fun findVector(rel: String): File {
    var d: File? = File(".").absoluteFile
    repeat(8) {
        val cur = d ?: throw ChanFailure("$rel not found")
        val p = File(cur, rel)
        if (p.isFile) return p
        d = cur.parentFile
    }
    throw ChanFailure("$rel not found from ${File(".").absolutePath}")
}

// ==== channel table vs the independent per-channel oracle (F3, non-circular) ========================

// The twenty channel directory names (lowercase), in TABLE registration order 0x0000..0x0013.
private val CHANNEL_DIRS = listOf(
    "control", "memory", "capability", "identity", "governance", "immune", "federation",
    "settlement", "compliance", "sensory", "telemetry", "audit", "stream", "bridge",
    "commerce", "interaction", "discovery", "workflow", "knowledge", "spatial",
)

private fun tableMatchesOracleRun() {
    for (id in 0..0x13) {
        val dir = CHANNEL_DIRS[id]
        val json = findVector("vectors/channels/$dir/cases.json").readText(Charsets.UTF_8)
        val ch = Channels.channel(id.toLong()) ?: throw ChanFailure("channel $dir (0x%04x)".format(id) + " not registered in Channels")

        katCheck("$dir channel_id == oracle", ch.id.toString(), intField(json, "channel_id").toString())
        katCheck("$dir name == oracle", ch.name, strField(json, "name"))

        val oracleKinds = objectArrayField(json, "kinds")
        katCheck("$dir kind count == oracle", ch.kinds.size.toString(), oracleKinds.size.toString())
        for ((i, kd) in ch.kinds.withIndex()) {
            val ok = oracleKinds[i]
            katCheck("$dir kind[$i].code == oracle", kd.code.toString(), intField(ok, "code").toString())
            katCheck("$dir kind[$i].name == oracle", kd.name, strField(ok, "name"))
            katCheck("$dir kind[$i].effect == oracle", kd.effect.toString(), intField(ok, "effect").toString())
            katCheckBool("$dir kind[$i].variable == oracle", kd.variable, boolField(ok, "variable"))
        }

        val oracleTransitions = objectArrayField(json, "transitions")
        katCheck("$dir transition count == oracle", ch.transitions.size.toString(), oracleTransitions.size.toString())
        for ((i, tr) in ch.transitions.withIndex()) {
            val ot = oracleTransitions[i]
            katCheck("$dir transition[$i].from == oracle", tr.first, strField(ot, "from"))
            katCheck("$dir transition[$i].to == oracle", tr.second, strField(ot, "to"))
        }

        katCheck("$dir states == oracle", ch.states.joinToString(","), stringArrayField(json, "states").joinToString(","))
        katCheck("$dir errors == oracle", ch.errors.joinToString(","), stringArrayField(json, "errors").joinToString(","))
    }
}

// ==== completeness: all twenty channels present, none thinned (R-11.1/R-11.2) ======================

private fun completenessRun() {
    for (id in 0..0x13) {
        val ch = Channels.channel(id.toLong())
        if (ch == null) {
            katFail("channel 0x%04x registered".format(id), "channel(id) returned null")
            continue
        }
        katOk("channel 0x%04x registered".format(id))
        katCheckBool("${ch.name} has at least one kind", ch.kinds.isNotEmpty(), true)
        for (kd in ch.kinds) {
            katCheckBool("${ch.name}.${kd.name} effect in range", kd.effect in 0..3, true)
        }
    }
    // an unregistered channel id returns null, mirroring Go channels.Channel.
    katCheckBool("unregistered channel 0x00FF returns null", Channels.channel(0x00FF) == null, true)
}

// ==== AllowedTransition: representative accept/reject cases (mirrors Go TestStateMachine) ==========

private fun stateMachineRun() {
    data class Case(val ch: Long, val from: String, val to: String, val ok: Boolean)
    val cases = listOf(
        Case(0x0001, "offered", "accepted", true),      // Memory
        Case(0x0001, "live", "revoked", true),           // Memory
        Case(0x0001, "revoked", "live", false),          // Memory regression
        Case(0x000E, "order", "fulfil", true),           // Commerce
        Case(0x000E, "offer", "fulfil", false),          // Commerce skip
        Case(0x0011, "awaiting-input", "running", true), // Workflow
        Case(0x0011, "created", "running", false),       // Workflow gate skip
    )
    for (c in cases) {
        val got = Channels.allowedTransition(c.ch, c.from, c.to)
        katCheckBool("channel 0x%04x %s->%s allowed".format(c.ch, c.from, c.to), got, c.ok)
    }
}

// ==== CheckFrameTree: valid tree accepted, cyclic tree rejected TransformCycle (Spatial §20) ========

private fun frameTreeRun() {
    val tree = mapOf("base" to "", "arm" to "base", "hand" to "arm")
    katExpectOk("valid frame tree accepted") { Channels.checkFrameTree(tree) }

    val cyclic = mapOf("a" to "b", "b" to "c", "c" to "a")
    katExpectThrows("cyclic frame tree rejected", "TransformCycle") { Channels.checkFrameTree(cyclic) }
}

// ==== Workflow input gate: the crash test that InputGateBypass cannot occur (design-channels §18) ==

private fun workflowGateInputBypassRun() {
    val wal = File.createTempFile("naalp-wf", ".wal").also { it.deleteOnExit() }
    wal.delete() // OpenWorkflowGate creates it; start from a genuinely-absent path like Go's t.TempDir()+"wf".

    val g = Channels.openWorkflowGate(wal.absolutePath)
    g.create("t1", false)

    // *** mutation anchor: running before input is supplied MUST be InputGateBypass. Neutering the
    // *** fail-closed branch in WorkflowGate.run flips this assertion from PASS to FAIL.
    katExpectThrows("task cannot run before its input gate", "InputGateBypass") { g.run("t1") }

    // Simulate a crash right after create(): close() flushes and closes the WAL; reopening recovers
    // the pre-gate durable status rather than bypassing it.
    g.close()
    val g2 = Channels.openWorkflowGate(wal.absolutePath)
    katCheck("crash recovers the pre-gate status", g2.status("t1") ?: "<null>", "awaiting-input")
    katExpectThrows("task cannot run after a crash without passing the gate", "InputGateBypass") { g2.run("t1") }

    // Supplying input opens the gate; only then may it run.
    katExpectOk("supplyInput opens the gate") { g2.supplyInput("t1") }
    katExpectOk("task runs after input is supplied") { g2.run("t1") }
    katCheck("status after run", g2.status("t1") ?: "<null>", "running")
    g2.close()

    wal.delete()
}

fun main() {
    println("ChannelsKatTest (Kotlin) — C10 channel state machine + Workflow input gate (design-channels.md)")
    tableMatchesOracleRun()
    completenessRun()
    stateMachineRun()
    frameTreeRun()
    workflowGateInputBypassRun()
    println(if (katFails == 0) "ChannelsKatTest: PASS" else "ChannelsKatTest: FAIL ($katFails)")
    if (katFails != 0) kotlin.system.exitProcess(1)
}
